src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.cs

```diff
@@ -0,0 +1,265 @@
+            nameof(Enumerable.Average),
+            nameof(Enumerable.Contains),
+            nameof(Enumerable.Count),
+            nameof(Enumerable.DefaultIfEmpty),
+            nameof(Enumerable.ElementAt),
+            nameof(Enumerable.ElementAtOrDefault),
+            nameof(Enumerable.First),
+            nameof(Enumerable.FirstOrDefault),
+            nameof(Enumerable.Last),
+            nameof(Enumerable.LastOrDefault),
+            nameof(Enumerable.LongCount),
+            nameof(Enumerable.Max),
+            nameof(Enumerable.Min),
+            nameof(Enumerable.Single),
+            nameof(Enumerable.SingleOrDefault),
+            nameof(Enumerable.Sum),
+            nameof(Enumerable.ToArray),
+            nameof(Enumerable.ToDictionary),
+            nameof(Enumerable.ToList),
+            nameof(Enumerable.ToLookup),
+            nameof(Enumerable.SequenceEqual),
+            // Only available on .net6 or later
+            "MaxBy",
+            "MinBy",
+            // Only available on .netstandard 2.1 or later
+            "ToHashSet");
+
+        /// <summary>
+        /// Linq methods deferring its parameters to be enumerated.
+        /// </summary>
+        public static readonly ImmutableArray<string> s_deferParametersEnumeratedLinqMethods = ImmutableArray.Create(
+            nameof(Enumerable.Append),
+            nameof(Enumerable.AsEnumerable),
+            nameof(Enumerable.Cast),
+            nameof(Enumerable.Distinct),
+            nameof(Enumerable.GroupBy),
+            nameof(Enumerable.OfType),
+            nameof(Enumerable.OrderBy),
+            nameof(Enumerable.OrderByDescending),
+            nameof(Enumerable.Prepend),
+            nameof(Enumerable.Reverse),
+            nameof(Enumerable.Select),
+            nameof(Enumerable.SelectMany),
+            nameof(Enumerable.Skip),
+            nameof(Enumerable.SkipWhile),
+            nameof(Enumerable.Take),
+            nameof(Enumerable.TakeWhile),
+            nameof(Enumerable.ThenBy),
+            nameof(Enumerable.ThenByDescending),
+            nameof(Enumerable.Where),
+            nameof(Enumerable.Concat),
+            nameof(Enumerable.Except),
+            nameof(Enumerable.GroupJoin),
+            nameof(Enumerable.Intersect),
+            nameof(Enumerable.Join),
+            nameof(Enumerable.Union),
+            nameof(Enumerable.Zip),
+            // Only available on .net6 or later
+            "Chunk",
+            "DistinctBy",
+            "ExceptBy",
+            "IntersectBy",
+            "UnionBy",
+            // Only available on .netstandard 2.1 or later
+            "TakeLast",
+            "SkipLast");
+
+        /// <summary>
+        /// Special Linq methods that no effect on its parameter, and not return new IEnumerable instance.
+        /// </summary>
+        private static readonly ImmutableArray<string> s_noEffectLinqChainMethods = ImmutableArray.Create(
+            nameof(Enumerable.AsEnumerable));
```

Comment:
Ah, this one is a special one. `s_deferParametersEnumeratedLinqMethods` should have it.

It is used to solve this case
```
var a = [|b|].AsEnumerable();
a.Count();
a.Count()
```
In this case, because [AsEnumerable](https://github.com/dotnet/runtime/blob/bdae17fc0ec222e75e651d96af74e9667c01602e/src/libraries/System.Linq/src/System/Linq/Enumerable.cs#L10) is not creating a new IEnumerable<T> (different from `Select`, `Where`.. they all create a new IEnumerable<> instance when called), so the diagnostic is reported at `b`.
But how it works is, the analyzer see `a.Count()`, then looks the PointToAnalzyerResult  of `a`, then it knows, 
`b.AsEnumerable()` create `a`. And because `AsEnumerable` is listed in the LinqChain, it will use a BFS to find 
`a` is actually a short-cut of `b`
The code is [here](https://github.com/dotnet/roslyn-analyzers/blob/927495b3992d289ba63be8c62d321954e53b8412/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs#L156)

Reply:
i see, then shouldn't it squiggle on the `a.Count()` instead of `[|b|]`?

Reply:
I would prefer to put it under `b` because in we flag the field/parameter all other cases